msgpack/ext.py

```diff
@@ -157,12 +157,14 @@ def to_datetime(self):
         :rtype: `datetime.datetime`
         """
         utc = datetime.timezone.utc
-        return datetime.datetime.fromtimestamp(0, utc) + datetime.timedelta(seconds=self.to_unix())
+        return datetime.datetime.fromtimestamp(0, utc) + datetime.timedelta(
+            seconds=self.seconds, microseconds=round(self.nanoseconds / 1e3)
```

Comment:
I prefer `//` over round.

Consider "12:34:56.78". It is "12:34" or "12:34:56", not "12:35" nor "12:34:57".

Reply:
> Just use 1000 instead of 1e3?

we can of course use 1000 instead, its just that in the code that i checked, i either saw "10**x" or "1ex" and wanted follow that convention.

> I prefer // over round.
> 
> Consider "12:34:56.78". It is "12:34" or "12:34:56", not "12:35" nor "12:34:57".

The idea I had with the round is when working with a timestamp with nanosecond precision "100,000,900" nanoseconds i thought it should round to "100,001" microseconds, whereas "100,000,100" nanoseconds should be "100,000" microseconds.

With the case "12:34:56.78", I'm not actually knowledgable whether it should "12:34:56" or "12:34:57", if rounding isn't the norm in these cases, we should definitely remove it and the tests related to it.

Reply:
> Just use 1000 instead of 1e3?

same as the previous comment, I just though "10**x" or "1ex" was the convention that was being used.

> I prefer // over round.
> 
> Consider "12:34:56.78". It is "12:34" or "12:34:56", not "12:35" nor "12:34:57".

I'm not actually knowledgeable about whether there should be a round or not. The idea I had with the round is when working with a timestamp with nanosecond precision, lets say, "100,000,900" nanoseconds, I thought it should round to "100,001" microseconds, whereas "100,000,100" nanoseconds should be "100,000" microseconds.

With the case "12:34:56.78", I'm not sure whether it should "12:34:56" or "12:34:57", if rounding isn't the norm in these cases, we should definitely remove it and the tests related to it.

Reply:
https://github.com/golang/go/blob/a63907808d14679c723e566cb83acc76fc8cafc2/src/time/time.go#L1214-L1244

Reply:
* https://doc.rust-lang.org/src/core/time.rs.html#360
* https://doc.rust-lang.org/src/core/time.rs.html#384

Reply:
https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/time/Instant.html#toEpochMilli()

> If this instant has greater than millisecond precision, then the conversion will drop any excess precision information as though the amount in nanoseconds was subject to integer division by one million.

Reply:
hmm i see, thanks a lot it makes more sense. Then i'll remove the rounding and tests related to it.

Reply:
2023-12-31 23:59:59.999999500 -- year 2023, date 2023-12-31.

After round half by micros:

2024-01-01 00:00:00.000000 -- year 2024, date 2024-01-01.

This is why most languages uses consistent "round down" for subseconds.
Consistency is important to avoid bugs. Every programmer should love consistency.

Reply:
yup of course, I didn't look at it like that when I was proposing the change. Again it makes more sense to round down instead as it would be more consistant as you said. Thanks for the explanation!